src/main/java/com/limechain/trie/structure/nibble/Nibbles.java

```diff
@@ -92,191 +84,148 @@ public boolean startsWith(Nibbles prefix) {
-        return nibbles.stream()
-                .map(Nibble::asLowerHexDigit)
-                .collect(
-                        Collector.of(
-                                StringBuilder::new,
-                                StringBuilder::append,
-                                StringBuilder::append,
-                                StringBuilder::toString
-                        )
-                );
-    }
-
-    @Override
-    public String toString() {
-        return this.toLowerHexString();
-    }
-
-    @Override
-    public Stream<Nibble> stream() {
-        return nibbles.stream();
-    }
-
-    @Override
-    public boolean equals(Object o) {
-        if (this == o) return true;
-        if (o == null || getClass() != o.getClass()) return false;
-        Nibbles nibbles1 = (Nibbles) o;
-        return this.nibbles.equals(nibbles1.nibbles);
-    }
-
-
-    @Override
-    public int hashCode() {
-        return Objects.hash(this.nibbles);
-    }
-
-    @NotNull
-    @Override
-    public Object[] toArray() {
-        return nibbles.toArray();
-    }
-
-    @NotNull
-    @Override
-    public <T> T[] toArray(@NotNull T[] a) {
-        return nibbles.toArray(a);
-    }
-
-    @Override
-    public boolean add(Nibble nibble) {
-        return nibbles.add(nibble);
-    }
-
-    @Override
-    public boolean remove(Object o) {
-        return nibbles.remove(o);
-    }
-
-    @Override
-    public boolean containsAll(@NotNull Collection<?> c) {
-        return new HashSet<>(nibbles).containsAll(c);
-    }
-
-    @Override
-    public boolean addAll(@NotNull Collection<? extends Nibble> c) {
-        return nibbles.addAll(c);
-    }
-
-    @Override
-    public boolean addAll(int index, @NotNull Collection<? extends Nibble> c) {
-        return nibbles.addAll(c);
-    }
-
-    @Override
-    public boolean removeAll(@NotNull Collection<?> c) {
-        return nibbles.removeAll(c);
+    private static String toLowerHexString(Iterable<Nibble> nibbles) {
+        return StreamSupport.stream(nibbles.spliterator(), false)
+            .map(Nibble::asLowerHexDigit)
+            .collect(
+                Collector.of(
+                    StringBuilder::new,
+                    StringBuilder::append,
+                    StringBuilder::append,
+                    StringBuilder::toString
+                )
+            );
     }
 
-    @Override
-    public boolean retainAll(@NotNull Collection<?> c) {
-        return nibbles.retainAll(c);
+    /**
+     * Adds a new nibble to the nibbles.
+     * @param nibble the new nibble to add
+     * @return the new Nibbles after insertion
+     * @implNote will create a new modified copy of the nibbles, does not mutate the instance invoked on
+     */
+    public Nibbles add(Nibble nibble) {
+        List<Nibble> newNibbles = new ArrayList<>(this.nibbles);
+        newNibbles.add(nibble);
+        return new Nibbles(newNibbles);
     }
```

Comment:
I don't like how mutable this method name sounds, but I decided it should be named after what it actually does and now after how it certainly behaves. Resolve if you don't take issue in that.